jwks.go

```diff
@@ -180,22 +191,84 @@ func (j *JWKS) RawJWKS() []byte {
 func (j *JWKS) ReadOnlyKeys() map[string]interface{} {
 	keys := make(map[string]interface{})
 	j.mux.Lock()
-	for kid, cryptoKey := range j.keys {
-		keys[kid] = cryptoKey.public
+	for _, key := range j.keys {
+		keys[key.kid] = key.Public
 	}
 	j.mux.Unlock()
 	return keys
 }
+func findKeyIdx(alg string, kid string, keys []ParsedJWK) int {
+	for idx, key := range keys {
+		if key.kid == kid {
+			return idx
+		}
+	}
+	return -1
+}
+
+func (j *JWKS) canUseKey(key ParsedJWK) bool {
+	canUseKey := true
+	// jwkUseWhitelist might be empty if the jwks was from keyfunc.NewJSON() or if JWKUseNoWhitelist option was true.
+	// in this case we don't restrict "use" parameter
+	if len(j.jwkUseWhitelist) > 0 {
+		_, canUseKey = j.jwkUseWhitelist[key.use]
+	}
+	return canUseKey
+}
+
+func (j *JWKS) getMatchingKeys(alg, kid, iss string) []*ParsedJWK {
+	result := make([]*ParsedJWK, 0, 1)
+	if kid != "" {
+		for idx, key := range j.keys {
+			if (key.kid == kid) && (key.algorithm == alg || key.algorithm == "") && j.canUseKey(key) {
+					result = append(result, &j.keys[idx])
```

Comment:
```suggestion
					result = append(result, key)
```
I thought keys were already pointers?
If not, the simpler thing is to also say: `key := key` and that breaks the loop aliasing.

Reply:
`key := key` looks nice, the reason for `&j.keys[idx]` was indeed loop aliasing.